pipelines/SpeciesTreeFromBusco/main.nf

```diff
@@ -292,18 +322,56 @@ process runIqtree {
     """
 }
 
+// Function to check if a genome is present in
+// the annotation cache.
+def annoInCache(genome, anno_cache) {
+    genome = new File("$genome")
+    genome = genome.getName()
+    File file = new File("${anno_cache}/${genome}/annotation.gtf")
+    return file.exists()
+}
+
 workflow {
     println(ensemblLogo())
+
+    // Prepare busco protein set:
     prepareBusco(params.busco_proteins)
+
+    // Prepare input genomes:
     genomes = Channel.fromPath("${params.dir}/*.*")
     prepareGenome(genomes)
 
-    buscoAnnot(prepareBusco.out.busco_prots, prepareGenome.out.proc_genome)
-    runGffread(buscoAnnot.out.busco_annot, buscoAnnot.out.genome)
+    // Branch the genomes channel based on presence in the
+    // annotation cache:
+    prepareGenome.out.proc_genome.branch {
+        annot: !annoInCache(it, params.anno_cache)
+        link: annoInCache(it, params.anno_cache)
+    }.set { genome_fork }
+
+    // Link annotations present in the cache:
+    linkAnnoCache(genome_fork.link)
+    // Annotate genomes not present in the cache:
+    buscoAnnot(prepareBusco.out.busco_prots, genome_fork.annot)
+
+    // Merge the output of link and annotate steps:
+    genomes = buscoAnnot.out.genome.mix(linkAnnoCache.out.genome)
+    annots = buscoAnnot.out.busco_annot.mix(linkAnnoCache.out.busco_annot)
```

Comment:
Do `genomes` and `annots` come out of the mix OK?

Reply:
Yes, this is a serious issue. Will have to pass around tuples to solve the potential mismatch in order.

Reply:
I understand this is to be addressed in an imminent pull request.